setup.py

```diff
@@ -1,75 +1,101 @@
-# Copyright 2018 D-Wave Systems Inc.
-#
-# Licensed under the Apache License, Version 2.0 (the "License");
-# you may not use this file except in compliance with the License.
-# You may obtain a copy of the License at
-#
-#   http://www.apache.org/licenses/LICENSE-2.0
-#
-# Unless required by applicable law or agreed to in writing, software
-# distributed under the License is distributed on an "AS IS" BASIS,
-# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
-# limitations under the License.
-
 import os
 from io import open
 from setuptools import setup, Extension
-from setuptools.command.build_py import build_py
+from setuptools.command.build_ext import build_ext
 
+class cythonizing_build_ext(build_ext):
+    """Cython extensions adaptive build: cythonize if possible, include numpy
+    headers, add custom compile/link flags."""
 
-# Load package info, without importing the package
-basedir = os.path.dirname(os.path.abspath(__file__))
-package_info_path = os.path.join(basedir, "tabu", "package_info.py")
-package_info = {}
-try:
-    with open(package_info_path, encoding='utf-8') as f:
-        exec(f.read(), package_info)
-except SyntaxError:
-    execfile(package_info_path, package_info)
+    # Custom extension compile/link flags
+    extra_compile_args = {
+        'msvc': ['/std:c++14'],
+        'unix': ['-std=c++11'],
+    }
 
+    extra_link_args = {
+        'msvc': [],
+        'unix': [],
+    }
 
-# Custom build command that runs `build_ext` before `build_py`, so that
-# SWIG-generated modules (in build_ext) are included in package (build_py).
-class build_ext_before_py(build_py):
-    def run(self):
-        self.run_command("build_ext")
-        return build_py.run(self)
+    def build_extensions(self):
+        # try using cython if available, fallback to using pre-cythonized
+        # files (if they exist; if not, ultimately fail)
+        try:
+            from Cython.Build import cythonize
 
+            # cythonization step will remove this flag needed by setuptools
```

Comment:
I am not sure that I understand what this step is doing

Reply:
My understanding is that Extension has an attribute called _needs_stub that is required by build_ext, but cythonize() removes it, so we have to save and restore it after cythonize().

Reply:
@arcondello, comes from https://github.com/dwavesystems/dwave-greedy/commit/3c9c506a7cabbc652b3b597fd440bed358eebc39.